### PR TITLE
Add write_verify support

### PR DESCRIPTION
**Pull request by bgly**
This commit adds WRITE VERIFY for qcow and file

Signed-off-by: Bryant G. Ly <bryantly@linux.vnet.ibm.com>

### CODE REVIEW

**Comment by bgly** [comment]
@agrover Review please

**Comment by mikechristie** [comment]
Could the do_verify_op and the write part of the implementations be made generic? It could look similar to vshankar's generic_compare_and_write where it takes the handler specific read/write functions as arguments.

**Comment by bgly** [comment]
I added a few more patches to this patch set, that addresses some other problems while testing bonnie++ performance test on user:file/user:file_async.

**Comment by bgly** [comment]
I have tested installing fedora using user:fbo onto a user:file backstore. Then I also manually created a file system using user:file_async/user:file backstore and ran bonnie++ performance tests on this patch without any failures.

**Comment by mikechristie** [comment]
I think the patches look ok.

Just 2 questions/comments.

1. Why was sync cache only implemented for the async file handler? Just wondering because it looked like you could support imm=0 for the sync case still.

2. Andy, is the policy going forward to do the generic_handle_cmd approach? It looks like some of those handlers have the out of bounds issue found and fixed in file_example here. Do we need the handle_cmd model? It looks like we could convert qcow and once we settle the debate on async vs sync we could convert file_example.

**Comment by bgly** [comment]
Yeah, I was just testing file_async so I bounded it towards that but it will also work for file sync. I uploaded a new patch

File: file_example.c [hunk 1]
```diff
@@ -360,33 +398,39 @@ static int file_handle_cmd(
...
+		unsigned int to_copy;
+		int i = 0;
 
 		remaining = length;
 
-		while (remaining) {
-			unsigned int to_copy;
-
-			to_copy = (remaining > iovec->iov_len) ? iovec->iov_len : remaining;
+		tcmu_info("Enter write: %m\n");
```
Comment (mikechristie) [inline file_example.c]: Hey.

Were the tcmu_info calls here and below supposed to be tcmu_dbg or deleted for the final merge?

File: main.c [hunk 2]
```diff
@@ -230,6 +230,9 @@ static int generic_handle_cmd(struct tcmu_device *dev,
 						   ASC_READ_ERROR, NULL);
 		} else
 			return SAM_STAT_GOOD;
+	case WRITE_VERIFY:
+		return store->write_and_verify(dev, tcmulib_cmd, iovec,
+					       iov_cnt, offset);
```
Comment (mikechristie) [inline main.c]: For this you mean to call the function you added right?

tcmu_emulate_write_verify()

?

File: tcmu-runner.h [hunk 3]
```diff
@@ -72,6 +72,8 @@ struct tcmur_handler {
 	ssize_t (*write)(struct tcmu_device *, struct iovec *, size_t, off_t);
 	ssize_t (*read)(struct tcmu_device *, struct iovec *, size_t, off_t);
 	int (*flush)(struct tcmu_device *);
+	int (*write_and_verify)(struct tcmu_device *, struct tcmulib_cmd *,
+				struct iovec *, size_t, off_t);
```
Comment (mikechristie) [inline tcmu-runner.h]: We can just drop this.

File: main.c (hunk 2 above)
Comment (bgly) [inline main.c]: Yes, fixed

File: tcmu-runner.h (hunk 3 above)
Comment (bgly) [inline tcmu-runner.h]: Done

File: file_example.c (hunk 1 above)
Comment (bgly) [inline file_example.c]: Deleted for final merge

Comment (bgly) [inline file_example.c]: Deleted for final merge

File: api.c [hunk 4]
```diff
@@ -905,3 +905,72 @@ int tcmu_emulate_start_stop(struct tcmu_device *dev, uint8_t *cdb,
...
+
+		if (rc != SAM_STAT_GOOD) {
+			tcmu_err("write_verify failed: %m\n");
+			return tcmu_set_sense_data(sense, MEDIUM_ERROR,
+						   ASC_READ_ERROR, NULL);
+
+		tcmu_seek_in_iovec(iovec, ret);
+		remaining -= ret;
+		}
+	}
```
Comment (mikechristie) [inline api.c]: Something got messed up in one of your uploads. That bracket was supposed to be before the tcmu_seek_in_iovec call right?

}
tcmu_seek_in_iovec(iovec, ret); 
remaining -= ret;

Do not worry about re-uploading. I will fix it up when I merge.

**Comment by mikechristie** [comment]
Merged through command line with these fixes:

https://github.com/open-iscsi/tcmu-runner/commit/5934dca954ad58400b574deb50528d2d69ba470b

**Pull request closed without merging**